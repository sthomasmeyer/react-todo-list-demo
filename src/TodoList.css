.TodoList {
  width: 666px;
  border: 5px solid lightskyblue;
  margin: 0 auto;
  align-items: center;
}

.TodoList h1 {
  color: midnightblue;
  text-align: center;
}
