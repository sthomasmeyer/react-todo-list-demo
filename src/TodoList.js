import React, { useState } from 'react';
import './TodoList.css';
import Todo from './Todo';
import NewToDoForm from './NewTodoForm';

const TodoList = () => {
  const [items, setItems] = useState([]);

  // This function is meant to be passed to the [NewTodoForm()] child of this...
  // [TodoList()] parent component.
  const addItem = (item) => {
    // The following [console.log()] commands are meant to illustrate the nature...
    // of this [addItem()] function, and the data that is being transmitted from...
    // child to parent.
    console.log(`unique id of to-do list item to be added: ${item.id}`);
    console.log(`text of to-do list item to be added: '${item.task}'`);

    // When this function is executed, it will update the current state of this...
    // 'TodoList' component based on the state of its child -- 'NewTodoForm'.
    setItems((items) => [...items, item]);
  };

  const updateItem = (id, updatedTask) => {
    setItems((items) =>
      items.map((item) =>
        item.id === id ? { ...item, task: updatedTask } : item
      )
    );
  };

  // This function is meant to be passed to the [Todo()] child of this...
  // [TodoList()] parent component.
  const removeItem = (id) => {
    setItems((items) => items.filter((item) => item.id !== id));
  };

  // Use JavaScript's [map()] method to cycle through each and every...
  // one of the item objects that is held in this component's current...
  // state (+) render a unique <div> component for each one.
  const todoItemComponents = items.map((item) => (
    <Todo
      key={item.id}
      id={item.id}
      task={item.task}
      updateItem={updateItem}
      removeItem={removeItem}
    />
  ));

  return (
    <div className='TodoList'>
      <h1>To Do List</h1>
      <NewToDoForm addItem={addItem} />
      <ul>{todoItemComponents}</ul>
    </div>
  );
};

export default TodoList;
