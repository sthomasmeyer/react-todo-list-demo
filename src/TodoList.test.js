import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

// Smoke test:
it('renders without crashing', () => {
  render(<TodoList />);
});

it('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it('generates a new item with an operational delete button', () => {
  render(<TodoList />);

  expect(screen.queryByText('X')).not.toBeInTheDocument();

  const addItemBtn = screen.queryByText('Add Task');

  fireEvent.click(addItemBtn);
  expect(screen.getByText('X')).toBeInTheDocument();

  const removeItemBtn = screen.getByText('X');

  fireEvent.click(removeItemBtn);
  expect(screen.queryByText('X')).not.toBeInTheDocument();
});

it('generates to-do list items based on user-input', () => {
  render(<TodoList />);

  function addItem(id = '2001', task = 'lori harvey') {
    const taskInput = screen.getByLabelText('Task:');
    fireEvent.change(taskInput, { target: { value: task } });
    const addItemBtn = screen.queryByText('Add Task');
    fireEvent.click(addItemBtn);
  }

  addItem();

  const addedTask = screen.queryByText('lori harvey');
  expect(addedTask).toBeInTheDocument();
});
