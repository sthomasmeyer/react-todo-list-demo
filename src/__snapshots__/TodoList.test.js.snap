// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1>
      To Do List
    </h1>
    <div
      class="NewTodoForm"
    >
      <form>
        <div>
          <label
            for="task"
          >
            Task: 
          </label>
          <input
            id="task"
            name="task"
            type="text"
            value=""
          />
        </div>
        <button>
          Add Task
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
